src/plugins/data/server/field_formats/field_formats_service.ts

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { has } from 'lodash';
+import { fieldFormats } from '../../common/field_formats';
+import { IUiSettingsClient } from '../../../../core/server';
+
+export class FieldFormatsService {
+  private readonly fieldFormatClasses: fieldFormats.IFieldFormatType[] =
+    fieldFormats.baseFormatters;
+
+  public setup() {
+    return {
+      register: (customFieldFormat: fieldFormats.IFieldFormatType) =>
+        this.fieldFormatClasses.push(customFieldFormat),
+    };
+  }
+
+  public start() {
+    return {
+      fieldFormatServiceFactory: async (uiSettings: IUiSettingsClient) => {
+        const fieldFormatRegistry = new fieldFormats.FieldFormatRegistry();
+        const uiConfigs = await uiSettings.getAll();
+        const registeredUiSettings = uiSettings.getRegistered();
+
+        Object.keys(registeredUiSettings).forEach(key => {
+          if (has(uiConfigs, key) && registeredUiSettings[key].type === 'json') {
+            uiConfigs[key] = JSON.parse(uiConfigs[key]);
+          }
+        });
+
+        fieldFormatRegistry.init((key: string) => uiConfigs[key], {}, this.fieldFormatClasses);
+
+        return fieldFormatRegistry;
+      },
+    };
+  }
+}
+
+/** @public */
+export type FieldFormatsSetup = ReturnType<FieldFormatsService['setup']>;
+
+/** @public */
+export type FieldFormatsStart = ReturnType<FieldFormatsService['start']>;
```

Comment:
nit: Can we make the way these types are defined consistent across the server & public field formats service? Because AFAICT they should be the same, right?

I'm wondering if perhaps we should export both interfaces from `common/field_formats` (one that's just register, one that excludes init/register), and then just continue to use `ReturnType` here (or explicitly define the interfaces; either way)

Reply:
This is IMO important

Reply:
@lukeelmers  @lizozom  I absolutely agree with you, but currently we have only one legal way to get `uiSettingsClient`, see `src/core/server/ui_settings/types.ts`

![image](https://user-images.githubusercontent.com/20072247/73162725-86c33680-40ff-11ea-80b7-b9029a421833.png)

This means that to create an instance of `uiSettingsClient` we need to get somehow the `RequestHandler` object.  Unfortunately, this is not possible from both the `setup` and `start` contracts.

I see no other way than to continue using `fieldFormatServiceFactory` 😞

Reply:
I think we can merge this PR and discuss this with @elastic/kibana-platform

Reply:
Ah, yes I misread that on the server the returned `fieldFormatsRegistry` is wrapped in the service factory. I thought they actually returned the same values which was why I commented in the first place.

In this case I agree with @lizozom, I don't want to block this PR on resolving it, so let's revisit later.

For now @alexwizp would you please add a `TODO` on the server service code noting that we should come back and remove the factory so server/client can have the same interface?